plugin/plugin_test.go

```diff
@@ -0,0 +1,134 @@
+package plugin
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+
+	"go.uber.org/multierr"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type mockParser struct {
+	flags                 []*flag
+	shouldErr             bool
+	errOnCallCount        int
+	addFlagGroupCallCount int
+}
+
+func (p *mockParser) AddFlagGroup(short, long string, data interface{}) error {
+	if p.shouldErr && p.errOnCallCount == p.addFlagGroupCallCount {
+		return errors.New("bad")
+	}
+	p.flags = append(p.flags, &flag{
+		shortDescription: short,
+		longDescription:  long,
+		data:             data,
+	})
+	p.addFlagGroupCallCount++
+	return nil
+}
+
+type fooFlags struct {
+	Bar string `long:"bar" description:"Sets the 'bar' value."`
+}
+
+// defer-able test cleanup for global state
+func setFlags(flags []*flag) (restore func()) {
+	_flags = flags
+	return func() {
+		_flags = nil
+	}
+}
+
+func TestAddToParser(t *testing.T) {
+	tests := []struct {
+		shouldErr                 bool
+		errOnCallCount            int
+		numErrs                   int
+		numAddFlagGroupCallCounts int
+		numParserFlags            int
+	}{
+		{numAddFlagGroupCallCounts: 3, numParserFlags: 3},
+		{shouldErr: true, numErrs: 3},
+		{shouldErr: true, errOnCallCount: 2, numErrs: 1, numAddFlagGroupCallCounts: 2, numParserFlags: 2},
+	}
+
+	threeFlags := []*flag{{}, {}, {}}
```

Comment:
This is still a very confusing test case. It may be easier to make it not-a-table-test.